server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java

```diff
@@ -0,0 +1,102 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.repositories;
+
+import org.elasticsearch.action.admin.cluster.repositories.get.GetRepositoriesResponse;
+import org.elasticsearch.action.admin.cluster.repositories.put.PutRepositoryResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.metadata.RepositoryMetaData;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.repositories.fs.FsRepository;
+import org.elasticsearch.snapshots.mockstore.MockRepository;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.test.InternalTestCluster;
+
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.sameInstance;
+
+public class RepositoriesServiceIT extends ESIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Collections.singletonList(MockRepository.Plugin.class);
+    }
+
+    public void testUpdateRepositoryWithDiffType() {
+        String repositoryName = "test-repo";
+
+        Client client = client();
+        InternalTestCluster cluster = (InternalTestCluster) cluster();
+        RepositoriesService repositoriesService = cluster.getInstance(RepositoriesService.class);
+        Settings settings = cluster.getDefaultSettings();
+        Path location = randomRepoPath();
+
+        Settings.Builder repoSettings = Settings.builder().put(settings).put("location", location);
+
+        PutRepositoryResponse putRepositoryResponse1 =
+            client.admin().cluster().preparePutRepository(repositoryName)
+                .setType(FsRepository.TYPE)
+                .setSettings(repoSettings)
+                .get();
+        assertThat(putRepositoryResponse1.isAcknowledged(), equalTo(true));
+
+        GetRepositoriesResponse getRepositoriesResponse1 =
+            client.admin().cluster().prepareGetRepositories(repositoryName).get();
+
+        assertThat(getRepositoriesResponse1.repositories(), hasSize(1));
+        RepositoryMetaData repositoryMetaData1 = getRepositoriesResponse1.repositories().get(0);
+
+        assertThat(repositoryMetaData1.type(), equalTo(FsRepository.TYPE));
+
+        Repository repository1 = repositoriesService.repository(repositoryName);
+        assertThat(repository1, instanceOf(FsRepository.class));
+
+        // update repository
+
+        boolean diffType = randomBoolean();
```

Comment:
Maybe we could make this more readable with:
`final String updatedRepositoryType = randomBoolean() ? "mock" : FsRepository.TYPE`

?